react/components/Metrics/Controllers/AppControllers/VersionInput.tsx

```diff
@@ -0,0 +1,81 @@
+import { any } from 'ramda'
+import React, { Component } from 'react'
+import { InjectedIntlProps, injectIntl } from 'react-intl'
+import * as semver from 'semver'
+import { Input } from 'vtex.styleguide'
+
+import { AppContext } from '../../Contexts/AppContext'
+
+
+interface State {
+  inputIsNotSemver: boolean
+}
+
+
+const semverRangesRegex = [
+  /([0-9])+\.x/,
+  /([0-9])+\.([0-9])+\.x/,
+  /\bv?(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-[\da-z-]+(?:\.[\da-z-]+)*)?(?:\+[\da-z-]+(?:\.[\da-z-]+)*)?\b/,
+]
+
+const isValidVersion = (version: string) => any((regex: RegExp) => regex.test(version), semverRangesRegex)
+
+class VersionInput extends Component<InjectedIntlProps, State> {
+  constructor(props: InjectedIntlProps) {
+    super(props)
+    this.state = {
+      inputIsNotSemver: false,
+    }
+  }
+
+  public setStateAsync(state: State) {
+    return new Promise((resolve) => {
+      this.setState(state, resolve)
+    })
+  }
+
+  public render = () => (
+    <AppContext.Consumer>
+      {({ appControllers, setAppControllers }) => (
+        <Input
+          placeholder="Ex: 1.0.x"
+          label={this.props.intl.formatMessage({ id: 'console.app.version' })}
+          onChange={(event: any) => this.parseInputAndSetState(event.target.value, appControllers, setAppControllers)}
+          errorMessage={this.state.inputIsNotSemver && 'Version is not in the SemVer spec'}
+        />
+      )}
+    </AppContext.Consumer>
+  )
+
+  private parseInputAndSetState = async (version: string, appControllers: AppController, setAppControllers: SetAppControllers) => {
+    const isValid = version.length === 0 || isValidVersion(version)
+    if (!isValid) {
+      await this.setStateAsync({ inputIsNotSemver: true })
+    } else {
+      await this.setStateAsync({ inputIsNotSemver: false })
```

Comment:
I don't think you should be doing this. React's `setState` [may batch or defer the update until later](https://reactjs.org/docs/react-component.html#setstate), so awaiting for a `setState` could lead to unwanted behaviour. What you could do is use its second argument, the callback, to do something after the new state is set. However, in this case, I believe you don't need to keep the validity of the input in the state, since it can be easily determined. I would move the `isValid` check to the `render` method, and send it as an argument to `parseInput`.

Reply:
I agree. I've tried your suggestion doing the following:

```ts
let inputIsSemver: boolean = true
return (
    <AppContext.Consumer>
        {({ appControllers, setAppControllers }) => (
        <Input
            placeholder="Ex: 1.0.x"
            label={this.props.intl.formatMessage({ id: 'console.app.version' })}
            onChange={(event: any) => {
            inputIsSemver = this.parseInput(event.target.value, appControllers, setAppControllers)
            
            }}
            errorMessage={inputIsSemver ? ''" : 'Version is not in the SemVer spec'}
        />
        )}
    </AppContext.Consumer>
)
```

but despite inputIsSemver is changing its value according to this.parseInput return, the errorMessage does not change. So, I decide to remove async/await and use the second argument callback of setState

Reply:
My suggestion was actually to move

```ts
const isValid = version.length === 0 || isValidVersion(version)
```

to the `render` method, but I just realized that `version` comes from the `onChange` `event`. Sorry.

In that case, I think your callback approach sounds good!